tests/checkpointing/test_legacy_checkpoints.py

```diff
@@ -86,3 +94,34 @@ def test_resume_legacy_checkpoints(tmpdir, pl_version: str):
     # assert result
 
     sys.path = orig_sys_paths
+
+
+class StatefulCallback(Callback):
+
+    def on_save_checkpoint(self, *args):
+        return {"content": 123}
+
+
+def test_callback_state_loading_by_type(tmpdir):
+    """ Test that legacy checkpoints that don't use a state identifier can still be loaded. """
+    model = BoringModel()
+    callback = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=1,
+        callbacks=[callback],
+    )
+    trainer.fit(model)
+    # simulate old format where type(callback) was the key
+    new_checkpoint = torch.load(Path(tmpdir, "last.ckpt"))
+    old_checkpiont = deepcopy(new_checkpoint)
+    old_checkpiont["callbacks"] = {type(callback): new_checkpoint["callbacks"]["ModelCheckpoint"]}
+    torch.save(old_checkpiont, Path(tmpdir, "old.ckpt"))
```

Comment:
Why don't you patch your ModelCheckpoint instance instead?

`callback.state_identifier = callback._legacy_state_identifier`

so you don't have to load and modify the checkpoint manually

Reply:
eehm yeah, this test wasn't good. improved now